### ISSUE TITLE
`Compiler.Commit(...)`

### ISSUE THREAD

**Issue by Tabaie** [ConsenSys/gnark#394]
Using Fiat-Shamir inside SNARKs can be prohibitive when the statement is large due to the large number of variables to be hashed. As per Alex, Azam and Olivier's paper, add a `Commit` function that provides a commitment to those variables efficiently, regardless of their visibility.

Suggested change:
```go
type Compiler interface {
  ⋮
  Commit(...Variable) Variable
}
```

Use:
```go
func (c *MyCircuit) Define(api frontend.API) {
  ⋮ // compute variables v1, v2, v3, v4 of any visibility
  transcript := fiatshamir.NewTranscript(api, someHash, "first", "second")
  commitment := api.Compiler().Commit(v1, v2, v3, v4)
  transcript.Bind("first", commitment)
  ⋮
}
```

The `Commit` function can only be called once during `Define`.

**Comment by ivokub** [comment]
I like the simple interface to commit the variables.

I think it makes sense that it is possible to call this method only once. Otherwise there may be strange behaviour (for example, when compiling in parallel and the `Commit` functions are called in non-deterministic orders etc.), although I think this can be solved (we order the variables by their order in the witness, for example). But in the context of composability, it may be difficult to use. So lets say there are different gadgets/sub-circuits which want to commit to the elements, but they cannot (because it may already be called or someone else is expecting to call it). One option would to have some kind of accumulator etc. and the accumulated variables would later be committed asynchronously, but then it would be difficult to obtain the commitment itself 🤔.

And, I guess the method should also return errors. Imo there may be several cases where an error can happen -- we build vanilla-R1CS/PLONK without commitments (where in which case we should manually compute the commitment, or have some kind of other fallback); the inputs are somehow invalid; method has already been called etc.

At some point I was thinking about having some `Defer`/`Finalise` method for the compiler which allow to register functions which are to be called when circuits `Define` method has returned. To extend this idea for the commitments, we could have something like:
```
type Compiler interface {
    Commit(in... Variable) error // adds in to be commited
    PostCommitment(finaliser func(api API, commitment Variable) error) error
    ...
}
```
Now, after the `Define` method of the circuit returns, we compute the commitment and call all `finaliser` functions (`PostCommitment` can be called multiple times) with the commitment and API, such that we can add further constraints/gates. If the constraint system doesn't support commitments, then `PostCommitment` returns an error which allows the caller to implement fallbacks if needed. In the finaliser method it is now invalid to call `Commit`.

But this may be overthinking it.

**Comment by Tabaie** [comment]
Thank you @ivokub for the detailed feedback! I'll try to address them paragraph by paragraph. Please lmk if I understood your concerns correctly.

### Paragraph 1:
Our reasoning to make the `Commit` function single-use was that:
1. We're trying not to deviate from the paper too much and minimize disruption of the existing `gnark` codebase.
2. It still suffices for most reasonable applications: If you have two independent proof systems in your circuit, you can make one commitment to the all the variables used in either statement and use the same commitment in the initialization of both your fiat-shamir transcripts. We should check with our theory-oriented friends, but I think this poses no threat (i.e. the security of Fiat-Shamir can never suffer from hashing _too many_ things.) The only case where more than one commitment would be needed is when making a proof _about_ an in-circuit proof. I think we can for now declare such use-cases out of scope.

So I think the use case with two independent proof would look like this:
```go
var statement10, statement11 frontend.Variable  // variables used in the statement of proof 1
var statement20, statement21 frontend.Variable  // variables used in the statement of proof 2

commitment := api.Compiler().Commit(statement10, statement11, statement20, statement21)

transcript1 := fiatshamir.NewTranscript(api, someHash, "challenge1", "challenge2")
transcript2 := fiatshamir.NewTranscript(api, someHash, "challenge1", "challenge2")

transcript1.Bind("challenge1", []frontend.Variable{protocol1UniqueIdentifier, commitment} )
transcript2.Bind("challenge1", []frontend.Variable{protocol2UniqueIdentifier, commitment} )
```

**Comment by Tabaie** [comment]
### Paragraph 2 (errors):
Fair; how about this:
```go
Commit(... Variable) (Variable, error)
```
### Paragraph 3+:
Could you explain the rationale for the callbacks a bit more? I don't think I quite understood it.

**Comment by ivokub** [comment]
> Thank you @ivokub for the detailed feedback! I'll try to address them paragraph by paragraph. Please lmk if I understood your concerns correctly.
> 
> ### Paragraph 1:
> Our reasoning to make the `Commit` function single-use was that:
> 
> 1. We're trying not to deviate from the paper too much and minimize disruption of the existing `gnark` codebase.
> 2. It still suffices for most reasonable applications: If you have two independent proof systems in your circuit, you can make one commitment to the all the variables used in either statement and use the same commitment in the initialization of both your fiat-shamir transcripts. We should check with our theory-oriented friends, but I think this poses no threat (i.e. the security of Fiat-Shamir can never suffer from hashing _too many_ things.) The only case where more than one commitment would be needed is when making a proof _about_ an in-circuit proof. I think we can for now declare such use-cases out of scope.
> 
> So I think the use case with two independent proof would look like this:
> 
> ```go
> var statement10, statement11 frontend.Variable  // variables used in the statement of proof 1
> var statement20, statement21 frontend.Variable  // variables used in the statement of proof 2
> 
> commitment := api.Compiler().Commit(statement10, statement11, statement20, statement21)
> 
> transcript1 := fiatshamir.NewTranscript(api, someHash, "challenge1", "challenge2")
> transcript2 := fiatshamir.NewTranscript(api, someHash, "challenge1", "challenge2")
> 
> transcript1.Bind("challenge1", []frontend.Variable{protocol1UniqueIdentifier, commitment} )
> transcript2.Bind("challenge1", []frontend.Variable{protocol2UniqueIdentifier, commitment} )
> ```

Yes, we could totally do it. But, lets say in the context of gadgets, where there are several gadgets where we want to use commitment. Then we would have to do something like this.

```go
func (c *Circuit) Define(api frontend.API) error {
   std1ToCommit := gadget1.DoStuffBeforeCommitment(api)
   std2ToCommit := gadget2.DoStuffBeforeCommitment(api)
   commitment := api.Compiler().Commit(append(std1ToCommit, std2ToCommit...)...)
   gadget1.DoStuffAfterCommitment(api, commitment)
   gadget2.DoStuffAfterCommitment(api, commitment)
   return nil
}
```
So we would need to know that the gadgets work in two phases and not forget to call them in correct order etc. so that stuff would work nicely. It gets even worse when gadgets call other gadgets etc.

It would be easier, if we could do something like this:
```go
func (c *Circuit) Define(api frontend.API) error {
   std1ToCommit := gadget1.DoStuff(api) // may want to use commitment!
   std2ToCommit := gadget2.DoStuff(api) // may also want use the commitment!
   return nil
}
```

But, currently, we wouldn't be able to do this as `Compiler.Commit` can be called only once. And we have to give as inputs all variables. And we cannot share easily the commitment between different gadgets.

So, in my proposal, if there would be something like this:
```go
type Compiler interface {
    Commit(frontend.Variable) error
    PostCommit(func (api API, commitment Variable) error) error
}
```
then within the gadgets (in every one of them!) we can do:
```go
func (g Gadget) DoStuff(api API) {
    // ...
    api.Compiler.Commit(var1)
   // ...
   api.Compiler.Commit(var2)
   //
   api.Compiler.PostCommit(g.DoStuffAfterCommitment)
}
```
So, we can share the commitment without having to pass back and forth all the variables we want to commit, the actual commitment and requiring the circuit developer to keep track if a gadget requires calling post-commit method etc.


But that is just something nice-to-have. I recon that for the very initial iteration it makes sense to have it as simple as possible and make it more complex as we need.

**Comment by Tabaie** [comment]
Ah yes I think you're right.
@gbotrel Something similar to this came up in our convo as well. I think we should go with Ivo's proposal. What do you think?

**Comment by gbotrel** [comment]
I ... don't like too much the "PostCommit" hook structure, I think the compiler should know (by order of calls) and return an error if trying to do something that don't make sense.
To be clear; is our functional need always "I want a single unique commitments for all the circuit" ? That is, in the case of multiple gadgets, don't we want to have independent commitments? 

Why not have an object that is similar to fiat shamir transcript; 
```
t := api.Compiler().NewCommitmentTranscript() // or some better name
x := api.Mul(y, z)
t.Commit(x)
// ... 
c := t.GetCommitment()
e := api.Mul(c, z)
// any subsequent call to t.Commit or t.GetCommitment panic, transcript has a state that's incoherent
```

If by design 2 gadgets/functions need to share the "transcript", then they do. If not they'll have 2 independent commitments. (or does it require more modifications in the Groth16 backend??)

**Comment by ivokub** [comment]
> I ... don't like too much the "PostCommit" hook structure, I think the compiler should know (by order of calls) and return an error if trying to do something that don't make sense. To be clear; is our functional need always "I want a single unique commitments for all the circuit" ? That is, in the case of multiple gadgets, don't we want to have independent commitments?
> 
> Why not have an object that is similar to fiat shamir transcript;
> 
> ```
> t := api.Compiler().NewCommitmentTranscript() // or some better name
> x := api.Mul(y, z)
> t.Commit(x)
> // ... 
> c := t.GetCommitment()
> e := api.Mul(c, z)
> // any subsequent call to t.Commit or t.GetCommitment panic, transcript has a state that's incoherent
> ```
> 
> If by design 2 gadgets/functions need to share the "transcript", then they do. If not they'll have 2 independent commitments. (or does it require more modifications in the Groth16 backend??)

I think Arya would know it better but I think that from the proof system perspective it is better to have only a single commitment. Otherwise the different commitment may be related to each other and the security properties wouldn't hold etc.
Even if we have only a single commitment, then this could be expanded to independent commitments through hashing. I guess the hashing can be done in- or out-circuit:
```
c <- Commit() // global unique single commitment built on top of all committed variables
c1, ..., cn <- Hash(1, c), ..., Hash(n, c)
return c1, ..., cn // to every invocation to t.GetCommitment()
```

**Comment by ivokub** [comment]
> I think security-wise a single commitment and multiple commitments are the same.

But, the same variable cannot be committed easily multiple times? It would be part of several sums and then we would have to have another element in the proof to cancel the multiples?

> Seems like the trade-off between single and multi commitment is performance vs elegance. With multi-commitments, the user could go:
> 
> ```go
> c1 := api.Compiler().Commit(v1, v2, v3)
> transcript1 := fiatshamir.NewTranscript(api, someHash, "first", "second")
> transcript1.Bind("first", c1)
> ⋮ // the rest of protocol 1, all in one place
> 
> c2 := api.Compiler().Commit(v3, v4, v5)  // both statements have v3 in common
> transcript2 := fiatshamir.NewTranscript(api, someHash, "first", "second")
> transcript2.Bind("first", c2)
> ⋮ // the rest of protocol 2, all in one place
> ```
I like the example. I think the main workflow is that the commitment is used to initialise Fiat-Shamir transcript, but even then there are cases where it is OK to use the commitment directly and not force the workflow.

Am I understanding correctly, that in this example (if we have the single-commitment approach), actually what are being committed to are `v1, v2, v3, v4, v5` (i.e. `v3` with a single multiplicity?).
> 
> , which is quite elegant. Performance-wise, in Groth16 for example, the proof size is then `3+n` where `n` is the number of commitments. On the verifier side also there is one pairing product check per commitment (that can probably be mitigated using linear combination tricks, but still.) The prover and verifier keys also grow linearly with the number of commitments.
> 
> These overheads seem significant for a proof system sold on the idea of constant proof size. That's why I personally gravitate towards the single-commitment solution.

Seems very reasonable. And I guess that even if we have somewhere in documentation that `commitment --> another group element in the proof`, then people would miss it and have very large proofs and expensive verifications.

**Comment by Tabaie** [comment]
> But, the same variable cannot be committed easily multiple times? It would be part of several sums and then we would have to have another element in the proof to cancel the multiples?

Ah, you're right. This complexity already is enough to disqualify multiple commitments imo.
Also, I think that proof size issue is a very serious one since it can directly negate the "S" in "SNARK", another serious downsize for multiple commitments.

In my opinion we should ditch multiple commitments, and the remaining question is, callback or no callback. I can still work on developing the backend stuff while we deliberate so no rush.
